tracer/test/test-applications/integrations/Samples.Wcf/Server/CalculatorService.cs

```diff
@@ -22,26 +24,49 @@ public async Task<double> ServerTaskAdd(double n1, double n2)
             return result;
         }
 
-        public IAsyncResult BeginServerAsyncAdd(double n1, double n2, AsyncCallback callback, object state)
+        public IAsyncResult BeginServerAsyncAdd(double n1, double n2, bool throwsException, bool synchronouslyCompletes, AsyncCallback callback, object state)
         {
-            LoggingHelper.WriteLineWithDate($"[Server] Received BeginServerAsyncAdd({n1},{n2})");
-            var tcs = new TaskCompletionSource<double>(state);
+            LoggingHelper.WriteLineWithDate($"[Server] Received BeginServerAsyncAdd({n1},{n2},{throwsException},{synchronouslyCompletes})");
 
-            var task = PerformAddWithDelay(n1, n2);
-            task.ContinueWith(t =>
+            var asyncResult = new SimpleAsyncResult<Tuple<double, bool>>(state);
+
+            if (synchronouslyCompletes)
+            {
+                if (throwsException)
+                {
+                    throw new FaultException("Something happened");
+                }
+
+                asyncResult.CompleteSynchronously(Tuple.Create(n1 + n2, false));
+                callback(asyncResult);
+            }
+            else
             {
-                tcs.SetResult(t.Result);
-                callback(tcs.Task);
-            });
+                var task = PerformAddWithDelay(n1, n2);
+                task.ContinueWith(
+                    t =>
+                    {
+                        asyncResult.Complete(Tuple.Create(t.Result, throwsException));
+                        callback(asyncResult);
+                    });
```

Comment:
Good lord, I know async/await isn't perfect, but then you remember how things used to be 😬